opnode/node/api.go

```diff
@@ -0,0 +1,274 @@
+package node
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/holiman/uint256"
+
+	"github.com/ethereum/go-ethereum/rpc"
+)
+
+type ErrorCode int
+
+const (
+	UnavailablePayload ErrorCode = -32001
+)
+
+type Bytes32 [32]byte
+
+func (b *Bytes32) UnmarshalJSON(text []byte) error {
+	return hexutil.UnmarshalFixedJSON(reflect.TypeOf(b), text, b[:])
+}
+
+func (b *Bytes32) UnmarshalText(text []byte) error {
+	return hexutil.UnmarshalFixedText("Bytes32", text, b[:])
+}
+
+func (b Bytes32) MarshalText() ([]byte, error) {
+	return hexutil.Bytes(b[:]).MarshalText()
+}
+
+func (b Bytes32) String() string {
+	return hexutil.Encode(b[:])
+}
+
+type Bytes256 [256]byte
+
+func (b *Bytes256) UnmarshalJSON(text []byte) error {
+	return hexutil.UnmarshalFixedJSON(reflect.TypeOf(b), text, b[:])
+}
+
+func (b *Bytes256) UnmarshalText(text []byte) error {
+	return hexutil.UnmarshalFixedText("Bytes32", text, b[:])
+}
+
+func (b Bytes256) MarshalText() ([]byte, error) {
+	return hexutil.Bytes(b[:]).MarshalText()
+}
+
+func (b Bytes256) String() string {
+	return hexutil.Encode(b[:])
+}
+
+type Uint64Quantity = hexutil.Uint64
+
+type BytesMax32 []byte
+
+func (b *BytesMax32) UnmarshalJSON(text []byte) error {
+	if len(text) > 64+2+2 { // account for delimiter "", and 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalJSON(text)
+}
+
+func (b *BytesMax32) UnmarshalText(text []byte) error {
+	if len(text) > 64+2 { // account for 0x prefix
+		return fmt.Errorf("input too long, expected at most 32 hex-encoded, 0x-prefixed, bytes: %x", text)
+	}
+	return (*hexutil.Bytes)(b).UnmarshalText(text)
+}
+
+func (b BytesMax32) MarshalText() ([]byte, error) {
+	return (hexutil.Bytes)(b).MarshalText()
+}
+
+func (b BytesMax32) String() string {
+	return hexutil.Encode(b)
+}
+
+type Uint256Quantity = uint256.Int
+
+type Data = hexutil.Bytes
+
+type PayloadID [8]byte
+
+func (b *PayloadID) UnmarshalJSON(text []byte) error {
+	return hexutil.UnmarshalFixedJSON(reflect.TypeOf(b), text, b[:])
+}
+
+func (b *PayloadID) UnmarshalText(text []byte) error {
+	return hexutil.UnmarshalFixedText("PayloadID", text, b[:])
+}
+
+func (b PayloadID) MarshalText() ([]byte, error) {
+	return hexutil.Bytes(b[:]).MarshalText()
+}
+
+func (b PayloadID) String() string {
+	return hexutil.Encode(b[:])
+}
+
+type ExecutionPayload struct {
```

Comment:
Are these types able to be imported from geth?

Reply:
Some of the API types are, but I was not happy with all of the types.

- To match the quantity naming in the spec, I aliased: `type Uint64Quantity = hexutil.Uint64`
- To properly enforce a maximum of 32 bytes as per the spec, I added `BytesMax32`
- To enforce fixed-size 256 byte logs-bloom, I added `Bytes256`
- To not deal with untyped payload IDs or byte slices, I created a type for it: `PayloadID`
- For hashes and addresses we just use the geth types (`common.Hash`, common.Address`)

TLDR: more strictly typed and closer to spec in a few cases.